### ISSUE TITLE
Issue 4: [ML] Llama 3.2 Engine & DPO Interface

### ISSUE THREAD

**Issue by daveAnalyst** [daveAnalyst/EngelBERT#6]
This is a crucial architectural distinction.

**Short Answer:** We are **NOT** abandoning Progressive Sovereignty. We are simply **sequencing** it.

To build a system that switches between models (e.g., "Use TinyLlama on CPU" -> "Use Gemma on GPU"), you first need to build the **system** that runs *one* model perfectly.

#### Issue 4: `[ML] Llama 3.2 Engine, DSPy Optimization & DPO Loop`
**Assignee:** 
**Priority:** P1
**Labels:** `ML`, `DSPy`, `Research`

### Context
**MAJOR ARCHITECTURE UPDATE:** We are utilizing a modern "SLM Stack" (Small Language Model).
1.  **Base:** Llama 3.2 1B (Quantized GGUF) via `llama-cpp-python` (Replaces NanoGPT/TinyLlama).
2.  **Prompting:** **DSPy** (Declarative Self-Improving Python) to programmatically optimize reasoning traces instead of brittle prompt engineering.
3.  **Alignment:** **DPO + QLoRA** to fine-tune the model on user `Save/Dismiss` signals.

#### 1. The Inference Engine (llama-cpp)
- Set up `llama-cpp-python` to load `Llama-3.2-1B-Instruct-GGUF`.
- **Constraint:** Must run on CPU with <2GB RAM usage.
- *Why:* This is the baseline for "Progressive Sovereignty."

#### 2. The DSPy Signature (The Logic)
- Implement a DSPy Signature for the Insight Task.
- *Code Draft:*
  ```python
  class InsightDiscovery(dspy.Signature):
      """Analyze two notes and determine if they share a non-obvious, valuable connection."""
      note_a = dspy.InputField()
      note_b = dspy.InputField()
      relevance_score = dspy.OutputField(desc="A float 0.0-1.0 score")
      reasoning = dspy.OutputField(desc="Short explanation of the link")
  ```
- **Task:** Write the `dspy.Module` that calls the Llama 3.2 model to execute this signature.

#### 3. The Alignment Loop (DPO + QLoRA)
- Research the `trl` (Transformer Reinforcement Learning) library integration with `peft`.
- Create a script `train_dpo.py` that takes the `rlhf_feedback_log` (Pairs of Chosen/Rejected insights) and updates the **LoRA adapters**.

### Definition of Done
- [ ] Llama 3.2 1B loads locally via Python.
- [ ] A DSPy module successfully returns a score and reasoning for a pair of notes.
- [ ] A DPO training script exists that can load a dummy dataset and perform one update step without crashing RAM.
```

---

**Comment by daveAnalyst** [comment]
on it currently
